### PR TITLE
Frontend

### PR DESCRIPTION
**Pull request by wik-z**
## Description

Added some frontend revisions, including:
* better error handling (highlighting specific inputs and columns, beautifying messages)
* database columns collapse when clicked
* download/copy to clipboard button
* more MySQL types
* adding options to enums

### CODE REVIEW

File: resources/assets/js/components/MigrationForm.vue [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+            <div class="col-md-6">
+                <h3>Columns</h3>
+                <div class="alert alert-info" v-if="columns.length === 0">
+                    Click the add column button below to get started
+                </div>
+                <div class="columns-list">
+                    <column
+                        v-for="(column, columnIndex) in columns"
+                        :column="column"
+                        :handleRemoveColumn="removeColumn.bind(this, columnIndex)"
```
Comment (samturrell) [inline resources/assets/js/components/MigrationForm.vue]: Props should be hyphenated.

Not keen on using `this` in templates. If this is just an event handler, use custom events.

File: resources/assets/js/components/MigrationForm.vue [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                        :index="columnIndex"
+                        :key="columnIndex">
+                    </column>
+                </div>
+                <div class="text-right">
+                    <button @click.prevent="addColumn" class="btn btn-default">
+                        <span class="fa fa-plus-circle"></span> Add Column
+                    </button>
+                    <button @click.prevent="sendColumns" class="btn btn-primary" :disabled="columns.length === 0">
```
Comment (samturrell) [inline resources/assets/js/components/MigrationForm.vue]: `!columns.length`

File: resources/assets/js/components/MigrationForm.vue [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            sendColumns() {
+                if (this.loading) {
+                    return;
+                }
+                // clear the code property first
+                this.code = "";
+                this.loading = true;
+
+                axios
+                    .post('api/generate', {
```
Comment (samturrell) [inline resources/assets/js/components/MigrationForm.vue]: Use repository pattern

File: resources/assets/js/modules/Global.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    constructor() {
+        $(document).on('domReady', () => {
+            this.invalidInputHandler();
+        });
+
+        $(document).on('reloadInvalidInputHandler', this.invalidInputHandler);
+    }
+
+    invalidInputHandler() {
+        $(".form-control").on("invalid", function () {
```
Comment (samturrell) [inline resources/assets/js/modules/Global.js]: Don't mix double and single quotes. Single quotes are the convention.

File: resources/assets/js/modules/Global.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+class Global {
+    constructor() {
+        $(document).on('domReady', () => {
```
Comment (samturrell) [inline resources/assets/js/modules/Global.js:3]: Not keen on any of this being in jQuery as it desyncs the virtual dom

File: resources/assets/sass/migration.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
```
Comment (samturrell) [inline resources/assets/sass/migration.php]: What is this and why is it in the sass folder?

File: resources/views/layout.blade.php [hunk 7]
```diff
@@ -7,6 +7,10 @@
         <link rel="stylesheet" href="{{ mix('css/app.css') }}">
         <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
         <title>Migration Generator</title>
+
+        <style>
```
Comment (samturrell) [inline resources/views/layout.blade.php]: ??

File: resources/assets/js/components/Column.vue [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                               v-model="column.is_foreign_key">
+                        Foreign Key
+                    </label>
+                </div>
+
+                <div v-if="column.is_foreign_key">
+                    <div class="column-foreign-key">
+                        <div class="form-group">
+                            <label>References:</label>
+                            <input type="text" class="form-control" name="references" placeholder="MySQL Table name"
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: Really not a fan of this formatting

File: resources/assets/js/components/Column.vue [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            </div>
+        </div>
+    </div>
+</template>
+
+<script>
+    import { columns, cascades } from '../modules/DatabaseConfig';
+
+    export default {
+        name: 'column',
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: Add line between

File: resources/assets/js/components/Column.vue [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+
+        data() {
+            return {
+                types: columns,
+                cascades: cascades,
+            };
+        },
+
+        computed: {
+            beautyName() {
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: Terrible variable name

File: resources/assets/js/components/Column.vue [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+                    const regexp = new RegExp(`columns.${this.index}.*`);
+                    const matchArray = key.match(regexp);
+
+                    return matchArray ? matchArray[0] : null;
+                });
+
+                if (columnHasError) {
+                    // we need to add 'with-error' class instead of 'has-error'
+                    // thanks for that, bootstrap
+                    $(this.$refs.column).addClass('with-error');
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: More jquery

File: resources/assets/js/components/Column.vue [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+            },
+
+            handleColumnOptionBackspace(event) {
+                // if there are no allowed values, ignore this event
+                if (!this.column.allowed_values) {
+                    return;
+                }
+
+                // if the pressed key isn't backspace, ignore this event
+                if(event.key !== "Backspace") {
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: `if (`

File: resources/assets/js/components/Column.vue [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+
+            removeColumnOption(index) {
+                this.column.allowed_values.splice(index, 1);
+            }
+        },
+
+        watch: {
+            'column': {
+                handler: function() {
+                    $(this.$refs.column).removeClass('with-error');
```
Comment (samturrell) [inline resources/assets/js/components/Column.vue]: More jquery? 😕

**Review by samturrell — CHANGES_REQUESTED** [review]

**Comment by wik-z** [comment]
Too many changes to be done on the errors handling, will re-submit when I get to it

**Pull request closed without merging**